src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,325 @@
+                    if (node.Children.Count == 0)
+                        continue;
+
+                    // Convert a rule of `a: (x | y | z)` into:
+                    // a: x
+                    //  | y
+                    //  | z;
+                    if (node.Children.Count == 1 && node.Children[0] is Field field && field.IsToken)
+                    {
+                        ProcessProductions(node, field.Kinds.Select(k =>
+                            new List<TreeTypeChild> { new Field { Type = SyntaxToken, Kinds = { k } } }).ToArray());
+                    }
+                    else
+                    {
+                        ProcessProductions(node, node.Children);
+                    }
+                }
+            }
+
+            // The grammar will bottom out with certain lexical productions.  Just emit a few empty
+            // productions in the grammar file indicating what's going on, and making it so that the
+            // g4 file is considered legal (i.e. no rule references names of rules that don't
+            // exist).
+
+            foreach (var kind in s_lexicalTokens)
+                _nameToProductions.Add(kind.ToString(), new List<Production> { new Production("/* see lexical specification */") });
+
+            return GenerateResult();
+        }
+
+        private void ProcessProductions(TreeType node, params List<TreeTypeChild>[] productions)
+            => _nameToProductions[node.Name].AddRange(productions.Select(p => ProcessChildren(p, delim: " ")));
+
+        private string GenerateResult()
+        {
+            // Keep track of the rules we've emitted.  Once we've emitted a rule once, no need to do
+            // it again, even if it's referenced by another rule.
+            var seen = new HashSet<string>();
+            var normalizedRules = new List<(string name, ImmutableArray<string> productions)>();
+
+            // Process each major section first, followed by the entire list. That way we process
+            // any rules not hit transitively from those sections.
+            var orderedRules = s_majorRules.Concat(
+                _nameToProductions.Keys.OrderBy(a => a, StringComparer.Ordinal));
+            foreach (var rule in orderedRules)
+                AddNormalizedRules(rule);
+
+            return
+@"// <auto-generated />
+grammar csharp;" + Join("", normalizedRules.Select(t => Generate(t.name, t.productions)));
+
+            void AddNormalizedRules(string name)
+            {
+                // Only consider the rule if it's the first time we're seeing it.
+                if (seen.Add(name))
+                {
+                    // Order the productions alphabetically for consistency and to keep us independent
+                    // from whatever ordering changes happen in Syntax.xml.
+                    var sorted = _nameToProductions[name].OrderBy(v => v.Text, StringComparer.Ordinal).ToImmutableArray();
+                    if (sorted.Length > 0)
```

Comment:
nit: this condition could be extended to encompass the rest of the block: `if (sorted.Length > 0) { ... ; var references = ...; foreach (references) ... }`

Reply:
Done!

Reply:
done!